### ISSUE TITLE
Migrate built-in keyboards -> dynamic keyboards

### ISSUE THREAD

**Issue by kmcnaught** [OptiKey/OptiKey#674]
We've talked before about turning all the built-in XAML keyboards into dynamic keyboards as per the new XML spec. Hopefully @BNewing is going to start looking into this :). I've created this issue to open a discussion about any particular concerns / requests / discussions along the way. I imagine we'll find a few limitations in the dynamic keyboards as we go along... 

Some initial thoughts from me:
- At some point we'll need to decide how we use localised strings in dynamic keys. I'm guessing this shouldn't be more complicated than agreeing a protocol for a `<Label>` to contain a localised string vs. a hard-coded string, either via an attribute, different element type or using special characters in the string itself. 

- Are there any performance considerations here? Would loading an XML dynamic keyboard (including various resource lookups for symbols & localised text potentially be noticeably slower than loading a built-in XAML keyboard? My opinion so far is that having XML versions of the keyboards is very valuable in itself, because it gives users an easy starting point for customising keyboards, and if performance becomes an issue we can always consider keeping the XAML for runtime if absolutely necessary. (i.e. let's not worry about this yet)

- There will almost certainly be some features currently in XAML that are not yet in dynamic keyboards. For instance the use of Converters for visual properties. Let's discuss them here when they come up - we might need to keep some select keyboards as XAML, or add features to the dynamic keyboard spec. 

- At some point we'll also need the discussion of where and how built-in keyboards get stored and loaded

- We'll need to think about tests - it's really important we can't accidentally break the built-in keyboards. At a minimum we'd need to test that all built-in keyboards (including all linked ones) are parsed without fail. It would be harder to test that all the individual key behaviours are correct, but maybe someone has some ideas?

@AdamRoden for visibility

**Comment by JuliusSweetland** [comment]
@kmcnaught @BNewing Great idea and very glad Beverly may be able to help (hi Beverly!). I wouldn't be too concerned about performance yet. We have a reasonably idea of performance from the suite of dynamic keyboards we already have and they seem to be fine. I would think the complexity is going to come from (as you say) localisation and other customisations/features which exist in hard coded keyboards, but are not yet supported in dynamic keyboards. Looking forward to this though - it would be a great step forward for the project.

**Comment by AdamRoden** [comment]
@kmcnaught 
FYI: I do have an undocumented feature that provides Dynamic Keys the ability to display values from Resources and Settings. Here is an example of how it can be used.

>\<Label>
{Resource:ADJUST_AMOUNT_IN_PIXELS}: {Setting:MoveAndResizeAdjustmentAmountInPixels}
\</Label>

**Comment by BNewing** [comment]
Hi @JuliusSweetland! I haven't forgotten about this! Life has been a bit busy lately, but hoping to get to this soon 🤞

**Comment by JuliusSweetland** [comment]
I completely understand! Thanks for the update.

**Comment by kmcnaught** [comment]
> 
> 
> @kmcnaught
> FYI: I do have an undocumented feature that provides Dynamic Keys the ability to display values from Resources and Settings. Here is an example of how it can be used.
> 
> > <Label>
> > {Resource:ADJUST_AMOUNT_IN_PIXELS}: {Setting:MoveAndResizeAdjustmentAmountInPixels}
> > </Label>

Nice! This is great, and something I've often wondered about when replicating some mouse actions (but never wanted enough to pursue!). It would be nice if this could tie in with #682 if that gets implemented, so you could have one key cycle through values of a variable and another key report the current value.

**Comment by kmcnaught** [comment]
Just a thought - as a starting point it might be nice to have the main menu as the first built-in dynamic keyboard, and if this got hooked up to user settings this would immediately allow users to tweak their optikey experience really easily.

**Comment by JuliusSweetland** [comment]
Yes I would agree with that. The menu is the main driver of a user's
experience, so being able to customise it would be really powerful.

**Comment by alexandre-mbm** [comment]
The Dynamic Keyboards feature needs to be better documented.

The new keyboard (dynamic) **should be identical** to early (built-in).

**Comment by AdamRoden** [comment]
Is there an update on this?

**Comment by JuliusSweetland** [comment]
Not from me I'm afraid

**Comment by BNewing** [comment]
Sorry for the silence from me. Realistically, don't have capacity to do this at present :(. I'm sorry it's taken me so long to say. I'm behind on another project I'm volunteering on, which I was hoping to have wrapped up in spring, but then life got in the way over summer, and now that's pretty far behind schedule and needs my focus. 

Thank you for your time @kmcnaught and for talking me through this. This is such an awesome project, and part of why it's taken me so long is that I was really reluctant to not be a part of it. I might have more time next year, but in the meantime, I will step back from this ticket in case anyone else wants to work on it

**Comment by kmcnaught** [comment]
Thanks for the update @BNewing

**Comment by alexandre-mbm** [comment]
Do you want an automatized method to convert them?
